Proper signal handling

s0undt3ch opened the pull request:
REVIEW CAREFULLY!!!!

s0undt3ch commented:
Whoops... some fixes required.

cachedout commented:
Yup, looks like we've still got some issues here. :]

s0undt3ch commented:
Yeah, the code changes regarding the master signals handling was not _compatible_ with making the master's `start()` method run on a separate process(as opposed to instantiating and running the master on a separate process). With the latest changes, it should actually run the tests now.

s0undt3ch commented:
The lint failures are unrelated to my changes...

s0undt3ch commented:
Err. Tests are still not running on Jenkins....

jtand commented:
Go Go Jenkins!

s0undt3ch commented:
Go Go Jenkins!

s0undt3ch commented:
Go Go Jenkins!

s0undt3ch commented:
Go Go Jenkins!

s0undt3ch committed:
Use salt.utils.process.MultiprocessingProcess

s0undt3ch committed:
Use salt.utils.process.MultiprocessingProcess

s0undt3ch commented:
Go Go Jenkins!

s0undt3ch commented:
Go Go Jenkins!

s0undt3ch commented:
Go Go Jenkins!

s0undt3ch commented:
Go Go Jenkins!

s0undt3ch commented:
Go Go Jenkins!

s0undt3ch committed:
Single quotes

s0undt3ch committed:
Name the test daemon processes for ps output

s0undt3ch committed:
Teach Python how to `deepcopy` `salt.crypt.AsyncAuth`

s0undt3ch committed:
Teach Python how to `deepcopy` `salt.transport.zeromq.AsyncZeroMQReqChannel`

s0undt3ch committed:
Use the deep copy memo

s0undt3ch committed:
Remove code that shouldn't have been committed.

s0undt3ch commented:
/cc @jacksontj can you please check the commits where I implemented the deep copy dunder and see if they don't make any sense or are bluntly wrong?

Whenever we try to deep copy an object which has an ioloop as an attribute it will fail because the ioloop contains a thread lock which is not deep copiable...

jacksontj commented on salt/crypt.py:
```diff
@@ -377,6 +378,19 @@ def __singleton_init__(self, opts, io_loop=None):
+        cls = self.__class__
+        result = cls.__new__(cls, copy.deepcopy(self.opts, memo), io_loop=None)
+        memo[id(self)] = result
+        for key, value in self.__dict__.items():
```
If you are going to do deepcopy of the value, I'm not certain you need .items() (as that will make a shallow copy as well).

jacksontj commented:
The code looks alright, although I'm curious under what circumstances do we deepcopy all of these channels? That seems like something we shouldn't be doing ;)

s0undt3ch commented:
https://jenkins.saltstack.com/job/salt-linode-cent5/2169/console the reason why those 2 commits were added...

On further investigation and extra debugging it all went down to the threading Lock that io loops have which can't be deep copied. So, either we don't store references to the io_loop's and just get the current one(if that does not break current functionality) or, we implement `__deepcopy__`...

s0undt3ch committed:
Don't create a shallow copy if we're going to deep copy it

s0undt3ch committed:
Don't create a shallow copy if we're going to deep copy it

s0undt3ch committed:
Name some more runtests daemon processes

s0undt3ch commented:
@jacksontj: To make it clear, you're stack context changes is what's causing all that copying.... :]

s0undt3ch commented:
*your

s0undt3ch committed:
Take into account a possible race condition when timing out futures

s0undt3ch commented:
To add more to the clarification, the offending instances were found in `__context__`.... should we sanitize the context in case we can't deep copy it instead?

s0undt3ch commented:
Go Go Jenkins!

s0undt3ch committed:
Only setup the multiprocessing logging listener process once

s0undt3ch committed:
Give the multiprocessing logging listener process a name for ps output

s0undt3ch committed:
Allow not passing a queue to setup_multiprocessing_logging

s0undt3ch committed:
Add a SignalHandlingMultiprocessingProcess

This will handle SIGINT/SIGTERM and exit the process

s0undt3ch committed:
Reset existing signals before starting SignalHandlingMultiprocessingProcess

s0undt3ch committed:
Simplify signal reseting

s0undt3ch committed:
Uniform exit messages

s0undt3ch committed:
Use SignalHandlingMultiprocessingProcess in the scheduler

s0undt3ch committed:
Shutdown multiprocessing logging before daemonizing and reconfigure it afterwards

s0undt3ch committed:
Switch to SignalHandlingMultiprocessingProcess

s0undt3ch committed:
Don't inherit signals

s0undt3ch committed:
Add SUBDEBUG/SUBWARNING from the multiprocessing log_to_stderr logger

s0undt3ch committed:
Name the KeepAlive minion multiprocessing process

s0undt3ch committed:
Remove unused import

s0undt3ch committed:
Shutdown multiprocessing logging before daemonizing and reconfigure it afterwards

s0undt3ch committed:
Switch to SignalHandlingMultiprocessingProcess

Additionally shutdown multiprocessing logging before daemonizing and
reconfigure it afterwards

s0undt3ch committed:
Clean up req server shutdown

s0undt3ch committed:
Switch to SignalHandlingMultiprocessingProcess

s0undt3ch committed:
Identify what appended the JID to the process name

s0undt3ch committed:
Identify what appended the JID to the process name. Exit the process!!!

s0undt3ch committed:
IOLoop knows how/when to stop on signals

s0undt3ch committed:
Destroy the ZMQ context as a multiprocessing.util.Finalize call

s0undt3ch committed:
Explicitly call `minion.destroy()`

s0undt3ch committed:
Explicitly delete schedule

s0undt3ch committed:
Include the class name

s0undt3ch committed:
Make the __MP_LOGGING_QUEUE_PROCESS a daemon process

s0undt3ch committed:
Fix multiprocessing logging setup

* Only create a multiprocessing Queue for logging in the MainProcess
* Only start and stop the multiprocessing logging listener in the MainProcess
* Only setup multiprocessing logging if not in the MainProcess
* And, as incredible as it may seem, we need to sleep after setting up multiprocessing logging for the new process or we'll end up in some kind of futex wait lock

s0undt3ch committed:
Explain why we need the time.sleep

s0undt3ch committed:
Fix globals SyntaxWarning's

s0undt3ch committed:
Lint fix

s0undt3ch committed:
Switch to SignalHandlingMultiprocessingProcess

s0undt3ch committed:
A little lint

s0undt3ch committed:
Reword the sleep comment with a better explanation

s0undt3ch committed:
Add missing future with_statement import

s0undt3ch commented:
The tests should still pass, I added some commits which ended in https://github.com/saltstack/salt/pull/28697 and did a hard-reset to this branch'es PR leaving it on the last know commit(which was passing).

s0undt3ch commented:
Go Go Jenkins!
